### PR TITLE
optimize janitor by splitting an expensive UPDATE SQL query

### PR DESCRIPTION
**Pull request by bertranddemiddelaer**
optimize janitor by splitting an expensive SQL query (the list
feeding the IN may contain millions of entries) into 2 quick
queries, using the fact that steps.id is autoincremented *and indexed*

## Contributor Checklist:

* [ ] I have updated the unit tests
* [ ] I have created a file in the `master/buildbot/newsfragments` directory (and read the `README.txt` in that directory)
* [ ] I have updated the appropriate documentation

### CODE REVIEW

File: master/buildbot/db/logs.py [hunk 1]
```diff
@@ -356,11 +356,21 @@ def thddeleteOldLogs(conn):
...
             # we do it first to avoid having UI discrepancy
+
+            # SELECT steps.id from steps WHERE steps.started_at < older_than_timestamp ORDER BY steps.id DESC LIMIT 1;
+            res = conn.execute(
+                sa.select([model.steps.c.id])
+                .where(model.steps.c.started_at < older_than_timestamp)
+                .order_by(model.steps.c.id.desc())
+                .limit(1)
+            )
+            stepid_max = res.fetchone()[0]
```
Comment (p12tic) [inline master/buildbot/db/logs.py]: Could we add a comment explaining this? Something like: "we utilize the fact that steps.id is autoincrement, thus step start times are effectively sorted and we only need to find the step at the upper bound of steps to update.

**Comment by tardyp** [comment]
Looks like the main mistake is that timestamps are not indexed.
I think this would help a lot of api usecases to get that indexed.
Waiting for that this looks good.

**Comment by bertranddemiddelaer** [comment]
> Looks like the main mistake is that timestamps are not indexed.
> I think this would help a lot of api usecases to get that indexed.
> Waiting for that this looks good.

This is still an issue, and adding an index did not help much in my case because the original subrequest fed the IN list with millions of items :( Changing the requests made the cost a matter of seconds where it was previously minutes (and a "show full processlist" in mysql console shown that the original UPDATE request spent nearly all its time in preparing status)

File: master/buildbot/db/logs.py (hunk 1 above)
Comment (bertranddemiddelaer) [reply master/buildbot/db/logs.py]: Will do :)

**Pull request merged**